src/audio/mixer.c

```diff
@@ -65,69 +65,164 @@ static void mix_n_s16(struct comp_dev *dev, struct audio_stream *sink,
 		      const struct audio_stream **sources, uint32_t num_sources,
 		      uint32_t frames)
 {
-	int16_t *src;
+	int16_t *src[PLATFORM_MAX_CHANNELS];
 	int16_t *dest;
 	int32_t val;
+	int nmax;
+	int n1;
+	int n2;
 	int i;
 	int j;
-	int channel;
-	uint32_t frag = 0;
-
-	for (i = 0; i < frames; i++) {
-		for (channel = 0; channel < sink->channels; channel++) {
-			val = 0;
-
-			for (j = 0; j < num_sources; j++) {
-				src = audio_stream_read_frag_s16(sources[j],
-								 frag);
-				val += *src;
+	int k;
+	int n;
+	int processed = 0;
+	int nch = sink->channels;
+	int samples = frames * nch;
+
+	dest = sink->w_ptr;
+	for (j = 0; j < num_sources; j++)
+		src[j] = sources[j]->r_ptr;
+
+	while (processed < samples) {
+		nmax = samples - processed;
+		n1 = audio_stream_bytes_without_wrap(sources[0], src[0]) >> 1; /* divide 2 */
+		n2 = audio_stream_bytes_without_wrap(sink, dest) >> 1;
+		for (i = 1; i < num_sources; i++) {
+			n = audio_stream_bytes_without_wrap(sources[i], src[i]) >> 1;
+			n1 = MIN(n, n1);
+		}
+		n = MIN(n1, n2);
+		n = MIN(n, nmax);
+		for (i = 0; i < n; i += nch) {
+			for (j = 0; j < nch; j++) {
+				val = *src[0];
+				src[0]++;
+				for (k = 1; k < num_sources; k++) {
```

Comment:
initialise `val = 0` and run the loop from 0

Reply:
This was inner loop optimization to save one operation. I'd prefer to keep this way unless your is faster (for from zero).